#  PROJECT 4 - ARTIFICIAL INTELLIGENCE IN SMART GRID 

NAME - **ROGER KEWIN SAMSON**


HAWK ID - **A20563057**


Mail ID: rkewin@hawk.iit.edu


**SPRING 2024**


Illinois Institute Of Technology

                                            Date: MARCH 28 , 2024




# Brief overview of the project:

The project focuses on applying these both datasets - Scikit-Learn Breast Cancer dataset, and the Scikit-Learn Diabetes dataset focusing on the Random Forest and Gradient Boosting algorithms. 

## DATA PREPARATION:
1. As per the given procedures first we split the dataset into training / validation (80%) and test (20%) using " train_test_split" library with a my own selected random state.
2. Then we prepare the second split from the training set from the 80% to training and validation sets, test size = 0.25 of it.




# RANDOM FOREST :

This method is used for both classification and regression task. it is used to correct the random forests for decision trees habit of overfitting to thier training set.
For classification, the output class is the one that receives the majority vote from individual trees. For regression, it is typically the average of the outputs from all trees.
The randomness injected into the model creation process helps prevent overfitting, making Random Forests more robust and accurate than individual decision trees.

# GRADIENT BOOSTING: 

It is the method used for both regression and classification tasks where to denote the weak prediction of models.Unlike Random Forest, Gradient Boosting builds trees one at a time, where each new tree helps to correct errors made by previously trained trees.

 It can be used with a variety of loss functions to model different prediction problems, making it a versatile tool for tackling a wide range of data science challenges.Both Random Forest and Gradient Boosting are powerful machine learning algorithms that can handle a wide range of data types and problem statements. While Random Forest aims to reduce variance through bagging, Gradient Boosting focuses on reducing bias through boosting, making them complementary approaches depending on the specific requirements

# 1(a) Python code that splits the original Wisconsin breast cancer dataset into two subsets: training/validation (80%), and test (20%). Be sure to document how you made the split, including the "random_state" value:

In [2]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

# Load the dataset
data = load_breast_cancer()
X = data.data
y = data.target

# Split the data into training/validation (80%) and test (20%) subsets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Print the shapes of the splits to verify the sizes
print("Training/Validation set size:", X_train.shape, y_train.shape)
print("Test set size:", X_test.shape, y_test.shape)




Training/Validation set size: (455, 30) (455,)
Test set size: (114, 30) (114,)


 In this above code we splits the wisconsin breast cancer dataset into two subsets by training/validation (80%)and test (20%)
we used the sklearn.diabetes and imported the datasets and train_test_split to split the data and get the output of the shape of  splitted datasets.

 

# I have used random_state=0. because, I want the consistent value for each repitation while running each time. So, I used "0" in the random state value in the above code

# 1(b) Python code that uses an additional split to create a validation dataset:

I have used the only additional split where the question has been seperated by "or". hence i used this method.

In [7]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

# Load the dataset
data = load_breast_cancer()
X = data.data
y = data.target

# First split: Separate the original dataset into 80% training/validation and 20% test
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Second split: Further split the training/validation dataset into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0) # This results in 60% training, 20% validation from the original dataset

# Output the sizes of each dataset to verify
print(f"Training set size: {X_train.shape}, {y_train.shape}")
print(f"Validation set size: {X_val.shape}, {y_val.shape}")
print(f"Test set size: {X_test.shape}, {y_test.shape}")

Training set size: (341, 30), (341,)
Validation set size: (114, 30), (114,)
Test set size: (114, 30), (114,)


I have created additional split as continued to the above code.The first split separate the original dataset into 80% training /validation  and 20% test with the random_state = 0. On the next split  60% training, 20% validation from the original dataset
with the same random_state=0.

# Procedure documenting your design process and the tradeoffs you considered in building a Random Forest Classifier.

As we must use the Random forest classifier it will handle both linear and nonlinear relationships.




## DATA PREPARATION :

•The required libraries are imported from the sklearn for this breast cancer data set used and assigned that datum in the X variable.
•The dataset is split using the library into training, validation, and testing. We already split it as 80% training/validation and 20% test and the training set is again split as a second split.

## HYPERPARAMETER SELECTION: 
•“n_estimators” (Number of trees) = 165: the decision to increase the number of trees from the default to improve model accuracy by averaging more decision trees. The tradeoff here is increased computational cost and potentially longer training times.
•‘max_depth = 10 ‘ : limiting the depth of trees helps to control the overfitting by preventing the model noisy patterns in training data.
•“min_samples_split=4”: Increasing the minimum number of samples required to split a node from  4 makes the model slightly more conservative in growing each tree. The trade-off involves potentially increased bias for slightly reduced variance.
•“random_state=0”: Setting a random state ensures that the result is reproducible.
Hence I used “0” to control the same randomness and get the same on each repetition to consider when looking for the best split at each node.

## MODEL EVALUATION :
To check the accuracy score by training, validation, and test datasets provide a straightforward measure of the model's performance. High training accuracy with comparable validation and test accuracies suggests good generalization.

# Python code that uses RandomForestClassifier to train, validate and test a Random Forest model. You may use any number of features from the dataset. Be sure to set the "random_state" so we can recreate your model.

In [4]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Load the dataset
data = load_breast_cancer()
X = data.data
y = data.target

# Split the dataset into training/validation (80%) and test (20%)
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Further split the training/validation dataset into training (75%) and validation (25%)
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Initialize the Random Forest Classifier with some hyperparameters adjusted from their default values
rf = RandomForestClassifier(n_estimators=165, max_depth=10, min_samples_split=4, random_state=0)

# Train the model on the training set
rf.fit(X_train, y_train)

# Evaluate the model on the training, validation, and test sets
y_pred_train = rf.predict(X_train)
y_pred_val = rf.predict(X_val)
y_pred_test = rf.predict(X_test)

# Calculate accuracy for each set
accuracy_train = accuracy_score(y_train, y_pred_train)
accuracy_val = accuracy_score(y_val, y_pred_val)
accuracy_test = accuracy_score(y_test, y_pred_test)

# Output the performance
print("RandomForestClassifier Performance:")
print(f"Training Accuracy: {accuracy_train:.4f}")
print(f"Validation Accuracy: {accuracy_val:.4f}")
print(f"Test Accuracy: {accuracy_test:.4f}")

RandomForestClassifier Performance:
Training Accuracy: 1.0000
Validation Accuracy: 0.9561
Test Accuracy: 0.9737


# What could you do to improve the prediction score of your trained model on the validation data?

•	Adjust ‘max_features’ new number of features to consider when looking for the best split. Reducing the max_features helps to improve the model's accuracy.

•	Increase ‘n_estimators’ the default value is 100 we increased the value is -165 still to improve the score increase the value. But it takes more computational time.

•	Use k-fold cross-validation to ensure the model's robustness across different subsets of the data. This can provide a more accurate estimate of the model's performance and generalization ability.

•	Use feature importance scores from the RandomForestClassifier to identify and retain only the most informative features. Removing irrelevant or noisy features can improve model accuracy.

•	Although Random Forest has built-in mechanisms to prevent overfitting, additional regularization techniques or more careful adjustment of the hyperparameters that control model complexity can further mitigate overfitting risks.


# How well does your final model predict the classes in the test data? Provide a list of all of the examples from the test data that are predicted incorrectly.

In [9]:
import numpy as np

# Identify a well-predicted example
correctly_predicted_indices = np.where(y_pred_test == y_test)[0]
if correctly_predicted_indices.size > 0:
    well_predicted_index = correctly_predicted_indices[0]
    print(f"Well-predicted example index: {well_predicted_index}")
    print(f"Predicted class: {y_pred_test[well_predicted_index]}, Actual class: {y_test[well_predicted_index]}")
else:
    print("No correctly predicted examples found.")

# Identify a poorly-predicted example
incorrectly_predicted_indices = np.where(y_pred_test != y_test)[0]
if incorrectly_predicted_indices.size > 0:
    poorly_predicted_index = incorrectly_predicted_indices[0]
    print(f"Poorly-predicted example index: {poorly_predicted_index}")
    print(f"Predicted class: {y_pred_test[poorly_predicted_index]}, Actual class: {y_test[poorly_predicted_index]}")
else:
    print("No incorrectly predicted examples found.")


No correctly predicted examples found.
Poorly-predicted example index: 0
Predicted class: 0.999999948577179, Actual class: 1


In [10]:


# Find the indices where predictions and actual labels differ
incorrect_indices = np.where(y_pred_test != y_test)[0]

# Print out the total number of incorrect predictions
print(f"Total Incorrect Predictions: {len(incorrect_indices)}")

# List all incorrectly predicted examples
print("Listing Incorrectly Predicted Examples:")
for index in incorrect_indices:
    print(f"Index: {index}, Predicted Class: {y_pred_test[index]}, Actual Class: {y_test[index]}")


Total Incorrect Predictions: 114
Listing Incorrectly Predicted Examples:
Index: 0, Predicted Class: 0.999999948577179, Actual Class: 1
Index: 1, Predicted Class: 8.526293151852991e-08, Actual Class: 0
Index: 2, Predicted Class: 8.526293146301876e-08, Actual Class: 0
Index: 3, Predicted Class: 0.999999948577179, Actual Class: 1
Index: 4, Predicted Class: 0.999999948577179, Actual Class: 1
Index: 5, Predicted Class: 8.526293151852991e-08, Actual Class: 0
Index: 6, Predicted Class: 8.526293151852991e-08, Actual Class: 0
Index: 7, Predicted Class: 0.19906166105411705, Actual Class: 0
Index: 8, Predicted Class: 0.35279422881945927, Actual Class: 1
Index: 9, Predicted Class: 0.9236645119118571, Actual Class: 1
Index: 10, Predicted Class: 0.6803999485771791, Actual Class: 1
Index: 11, Predicted Class: 8.526293146301876e-08, Actual Class: 0
Index: 12, Predicted Class: 1.0069133077481627, Actual Class: 1
Index: 13, Predicted Class: 0.08100006087268094, Actual Class: 0
Index: 14, Predicted Class

# Gradient Boosting learning algorithm. Reuse the same splits by using the same random_state values. This way, your Gradient Boosting Classifier will see the same training, validation and test data. 

In [13]:
# Import necessary libraries
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score

# Load the dataset
data = load_breast_cancer()
X = data.data
y = data.target

# Split the dataset into training/validation (80%) and test (20%) with a specific random_state
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Further split the training/validation dataset into training (75%) and validation (25%) with the same random_state
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Initialize and train the GradientBoostingClassifier with adjusted hyperparameters
gb_classifier = GradientBoostingClassifier(n_estimators=150, max_depth=10, min_samples_split=4, random_state=0)
gb_classifier.fit(X_train, y_train)

# Make predictions with the GradientBoostingClassifier
y_pred_train_gb = gb_classifier.predict(X_train)
y_pred_val_gb = gb_classifier.predict(X_val)
y_pred_test_gb = gb_classifier.predict(X_test)

# Calculate and output the performance
print("Gradient Boosting Classifier Performance:")
print(f"Training Accuracy: {accuracy_score(y_train, y_pred_train_gb):.4f}")
print(f"Validation Accuracy: {accuracy_score(y_val, y_pred_val_gb):.4f}")
print(f"Test Accuracy: {accuracy_score(y_test, y_pred_test_gb):.4f}")


Gradient Boosting Classifier Performance:
Training Accuracy: 1.0000
Validation Accuracy: 0.9123
Test Accuracy: 0.9298


In [33]:

# Here's the crucial step: Make predictions with the GradientBoostingClassifier
y_pred_test_gb = gb_classifier.predict(X_test)

# Now, perform the analysis
# Identify a well-predicted example
correctly_predicted_indices = np.where(y_pred_test_gb == y_test)[0]
if correctly_predicted_indices.size > 0:
    well_predicted_index = correctly_predicted_indices[0]
    print(f"Well-predicted example index: {well_predicted_index}")
    print(f"Predicted class: {y_pred_test_gb[well_predicted_index]}, Actual class: {y_test[well_predicted_index]}")
else:
    print("No correctly predicted examples found.")

# Identify a poorly-predicted example
incorrectly_predicted_indices = np.where(y_pred_test_gb != y_test)[0]
if incorrectly_predicted_indices.size > 0:
    poorly_predicted_index = incorrectly_predicted_indices[0]
    print(f"Poorly-predicted example index: {poorly_predicted_index}")
    print(f"Predicted class: {y_pred_test_gb[poorly_predicted_index]}, Actual class: {y_test[poorly_predicted_index]}")
else:
    print("No incorrectly predicted examples found.")


Well-predicted example index: 0
Predicted class: 0, Actual class: 0
Poorly-predicted example index: 10
Predicted class: 0, Actual class: 1


In [34]:
import numpy as np

# Assuming y_pred_test_gb contains your test predictions from the Gradient Boosting Classifier

# Find the indices where predictions and actual labels differ
incorrect_indices = np.where(y_pred_test_gb != y_test)[0]

# Print out the total number of incorrect predictions
print(f"Total Incorrect Predictions: {len(incorrect_indices)}")

# List all incorrectly predicted examples
print("Listing Incorrectly Predicted Examples:")
for index in incorrect_indices:
    print(f"Index: {index}, Predicted Class: {y_pred_test_gb[index]}, Actual Class: {y_test[index]}")


Total Incorrect Predictions: 8
Listing Incorrectly Predicted Examples:
Index: 10, Predicted Class: 0, Actual Class: 1
Index: 13, Predicted Class: 0, Actual Class: 1
Index: 31, Predicted Class: 1, Actual Class: 0
Index: 47, Predicted Class: 0, Actual Class: 1
Index: 91, Predicted Class: 1, Actual Class: 0
Index: 92, Predicted Class: 0, Actual Class: 1
Index: 97, Predicted Class: 0, Actual Class: 1
Index: 109, Predicted Class: 1, Actual Class: 0


both algorithms are highly effective, they do not perform identically across the dataset, as evidenced by the difference in their test accuracies. Since the Random Forest Classifier has a higher test accuracy (0.9737) compared to the Gradient Boosting Classifier (0.9298), it suggests that RFC is making fewer mistakes on the test data than GBC.

Without a detailed comparison of the instances each model predicts incorrectly, we cannot conclusively say if the mistakes overlap exactly

## Is One Clearly Better Than the Other?

Higher Test Accuracy: RFC's test accuracy of 0.9737 is superior to GBC's 0.9298. This indicates that RFC is more accurate in predicting unseen data.

Validation Accuracy: Although RFC also shows a higher validation accuracy (0.9561) compared to GBC (0.9123), indicating better generalization from the training to unseen data during the model tuning phase.

# ADVANTAGES & DISADVANTAGES

## Random Forest Classifier:
Advantages:

Robustness: RFC is less prone to overfitting due to its ensemble nature, making it reliable even when dealing with complex datasets.
Feature Insight: The ability to assess feature importance offers valuable insights into the dataset, aiding in feature selection and understanding the factors influencing predictions.


Disadvantages:

Computational Resource: With a large number of trees, RFC can become computationally demanding, impacting performance and scalability.
Performance Plateau: After a certain point, adding more trees yields diminishing returns on model improvement.


## Gradient Boosting Classifier:
Advantages:
Precision: GBC often achieves higher accuracy by focusing on correcting previous mistakes, making it effective for complex datasets.
Flexibility: Its ability to optimize different loss functions makes it versatile across a range of predictive modeling tasks.


Disadvantages:
Overfitting Risk: GBC can overfit noisy datasets due to its sequential, error-correcting approach, requiring careful tuning of its parameters.

Computational Intensity: The sequential nature of boosting makes GBC more computationally intensive than RFC, especially for large datasets.

# CONCLUSION


Both Gradient boosting classifier and Random forrest classifier have shown effectiveness in classifying the Wisconsin Breast Cancer dataset, each with distinct operational characteristics and implications for model selection.

Selection Criteria: The choice between them should consider the specific needs of the application, including accuracy requirements, computational resources, and model interpretability. For instance, RFC might be preferred for its robustness and lower computational cost, especially in scenarios where interpretability and quick model training are crucial. Conversely, GBC could be the choice for applications where the highest possible accuracy is paramount, and there are sufficient resources to fine-tune the model and manage its greater computational demands.

# DIABETES DATASET:

# 2(a).Python code that splits the original Diabetes dataset into two subsets: training/validation (80%), and test (20%). Be sure to document how you made the split, including the "random_state" value used in the shuffling process, so we can recreate your exact splits. See "model_selection.train_test_split" for guidance.

In [7]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split

# Load the dataset
diabetes_data = load_diabetes()
X = diabetes_data.data
y = diabetes_data.target

# Split the dataset into training/validation and test subsets
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Display the sizes of the splits
print(f"Training/Validation set size: {X_train_val.shape}, {y_train_val.shape}")
print(f"Test set size: {X_test.shape}, {y_test.shape}")


Training/Validation set size: (353, 10), (353,)
Test set size: (89, 10), (89,)


In this above code we splits the wisconsin breast cancer dataset into two subsets by training/validation (80%)and test (20%) we used the sklearn.diabetes and imported the datasets and train_test_split to split the data and get the output of the shape of splitted datasets.

And I used the random_state = 0 to get the same value while looping on each time.

# 2(b).Python code that uses an additional split to create a validation dataset.

In [9]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split

# Load the Diabetes dataset
diabetes_data = load_diabetes()
X = diabetes_data.data  # Features
y = diabetes_data.target  # Target variable

# Initial split: Separate the dataset into 80% training/validation and 20% test
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Further split the training/validation dataset into 75% training and 25% validation
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Display the sizes of each dataset to verify the splits
print("Final Training set size:", X_train.shape, y_train.shape)
print("Validation set size:", X_val.shape, y_val.shape)
print("Test set size:", X_test.shape, y_test.shape)



Final Training set size: (264, 10) (264,)
Validation set size: (89, 10) (89,)
Test set size: (89, 10) (89,)


I have created additional split as continued to the above code.The first split separate the original dataset into 80% training /validation and 20% test with the random_state = 0. On the next split 60% training, 20% validation from the original dataset with the same random_state=0.

# Procedure documenting your design process and the tradeoffs you considered in building a Random Forest Regressor.


•	Random Forest can handle a large number of features, but not all features may be relevant. The trade-off here is between model complexity and performance

•	The dataset has been split as  80% training /validation and 20% test with the random_state = 0. On the next split 60% training, 20% validation

•	Number of Trees (n_estimators): More trees can reduce variance and improve the model's accuracy but also increase computational time and complexity. Finding the right balance is key.

•	Tree Depth (max_depth): Deeper trees can model complex relationships but risk overfitting. 

•	Minimum Samples per Leaf (min_samples_leaf): This parameter helps control overfitting by defining how specific the model can be about classifying observations. 

•	Feature Sampling (max_features): Random Forest randomly selects a subset of features for splitting nodes. 

•	Cross-validation: Employing cross-validation techniques helps estimate the model's performance on unseen data and mitigate overfitting. The trade-off is computational cost, especially with a large dataset or numerous hyperparameters.

•	Hyperparameter Tuning: Utilizing methods like Grid Search or Random Search to find the optimal set of hyperparameters involves computational cost versus performance improvement trade-offs.

•	t's important to compare these metrics on both validation and test datasets to ensure the model hasn't overfitted.

•	Accuracy vs. Interpretability: Random Forest offers high accuracy but can be less interpretable than simpler models due to its ensemble nature.


# Python code that uses RandomForestRegressor to train, validate and test a Random Forest model. You may use any number of features from the dataset. Be sure to set the "random_state" so we can recreate your model.

In [20]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import numpy as np

# Load the Diabetes dataset
diabetes_data = load_diabetes()
X, y = diabetes_data.data, diabetes_data.target

# Split the dataset into training/validation (80%) and test (20%)
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Further split the training/validation dataset into training (75%) and validation (25%)
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Initialize RandomForestRegressor with modified hyperparameters
regressor = RandomForestRegressor(n_estimators=250, max_depth=10, min_samples_split=3, random_state=0)

# Train the model
regressor.fit(X_train, y_train)

# Evaluate the model
y_train_pred = regressor.predict(X_train)
y_val_pred = regressor.predict(X_val)
y_test_pred = regressor.predict(X_test)

# Calculate and print the scores
print("Training Score:", regressor.score(X_train, y_train))
print("Validation Score:", regressor.score(X_val, y_val))
print("Test Score:", regressor.score(X_test, y_test))



Training Score: 0.914734165625094
Validation Score: 0.3878068799999501
Test Score: 0.37464244602867913


## What could you do to improve the prediction score of your trained model on the validation data? How well does your final model predict the targets in the test data?


Conduct a more thorough hyperparameter tuning using GridSearchCV or RandomizedSearchCV. Explore beyond n_estimators, max_depth, and min_samples_split; consider max_features, min_samples_leaf, and max_leaf_nodes for potentially better performance.


Investigate the dataset for potential new features that could be engineered from the existing data. Sometimes, interactions between features or polynomial features can provide additional predictive power.
Use RandomForestRegressor's feature importance to identify and possibly eliminate features that contribute little to the model's predictive ability, thereby focusing the model on more relevant signals.

Implement k-fold cross-validation to ensure that the model's performance is consistent across different subsets of the data. This can help in identifying overfitting issues early and provide a more stable estimate of the model's performance.

In [21]:
# Calculate MSE and RMSE for the test data
mse_test = mean_squared_error(y_test, y_test_pred)
rmse_test = np.sqrt(mse_test)

print(f"Test MSE: {mse_test:.2f}")
print(f"Test RMSE: {rmse_test:.2f}")


Test MSE: 3206.79
Test RMSE: 56.63


Accuracy Evaluation: Utilize the model's R² score, Mean Squared Error (MSE), or Mean Absolute Error (MAE) on the test set to evaluate how well it predicts the targets. A higher R² score and lower MSE/MAE values indicate better model performance.

Consistency Across Datasets: Compare the model's performance across the training, validation, and test sets. Performance consistency suggests good generalization, whereas significant drops in test performance might indicate overfitting to the training data.

Error Analysis: Identifying the top 10 prediction errors on the test data, as previously discussed, can provide specific insights into instances where the model struggles, highlighting potential areas for model or data improvement.

## Provide a list of the top 10 prediction errors based on the examples from the test data.


In [22]:
# Assuming model is trained and y_test_pred is obtained
errors = np.abs(y_test_pred - y_test)
top_10_errors_indices = np.argsort(errors)[-10:]
print("Top 10 Prediction Errors:")
for index in top_10_errors_indices:
    print(f"Index: {index}, Predicted: {y_test_pred[index]:.2f}, Actual: {y_test[index]}, Error: {errors[index]:.2f}")


Top 10 Prediction Errors:
Index: 35, Predicted: 186.95, Actual: 281.0, Error: 94.05
Index: 31, Predicted: 82.70, Actual: 181.0, Error: 98.30
Index: 37, Predicted: 155.22, Actual: 257.0, Error: 101.78
Index: 67, Predicted: 205.13, Actual: 317.0, Error: 111.87
Index: 9, Predicted: 213.44, Actual: 99.0, Error: 114.44
Index: 58, Predicted: 115.31, Actual: 230.0, Error: 114.69
Index: 24, Predicted: 183.27, Actual: 68.0, Error: 115.27
Index: 65, Predicted: 183.94, Actual: 302.0, Error: 118.06
Index: 21, Predicted: 146.54, Actual: 276.0, Error: 129.46
Index: 60, Predicted: 188.34, Actual: 52.0, Error: 136.34


# Gradient Boosting learning algorithm. Reuse the same splits by using the same random_state values. This way, your Gradient Boosting Regressor will see the same training, validation and test data. When you have completed the process with the Gradient Boosting Regressor.

In [24]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error
import numpy as np


data = load_diabetes()
X, y = data.data, data.target

# Split the dataset into training/validation (80%) and test (20%)
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Further split the training/validation dataset into training (75%) and validation (25%)
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Initialize the GradientBoostingRegressor with specified hyperparameters
gb_regressor = GradientBoostingRegressor(n_estimators=1, max_depth=10, min_samples_split=4, random_state=0)

# Train the GradientBoostingRegressor on the training set
gb_regressor.fit(X_train, y_train)

# Make predictions on the training, validation, and test sets
y_pred_train = gb_regressor.predict(X_train)
y_pred_val = gb_regressor.predict(X_val)
y_pred_test = gb_regressor.predict(X_test)

# Evaluate the model's performance
train_score = gb_regressor.score(X_train, y_train)
val_score = gb_regressor.score(X_val, y_val)
test_score = gb_regressor.score(X_test, y_test)

# Output the performance scores
print("Gradient Boosting Regressor Performance:")
print(f"Training Score: {train_score:.4f}")
print(f"Validation Score: {val_score:.4f}")
print(f"Test Score: {test_score:.4f}")

# Identify and output the top 10 prediction errors on the test set
errors = np.abs(y_test - y_pred_test)
top_10_errors_indices = np.argsort(errors)[-10:]

print("\nTop 10 Prediction Errors (Indices, Actual Values, Predicted Values):")
for idx in top_10_errors_indices:
    print(f"Index: {idx}, Actual: {y_test[idx]}, Predicted: {y_pred_test[idx]:.2f}")



Gradient Boosting Regressor Performance:
Training Score: 0.1801
Validation Score: 0.0460
Test Score: 0.0739

Top 10 Prediction Errors (Indices, Actual Values, Predicted Values):
Index: 10, Actual: 252.0, Predicted: 143.23
Index: 37, Actual: 257.0, Predicted: 145.13
Index: 35, Actual: 281.0, Predicted: 154.56
Index: 5, Actual: 275.0, Predicted: 145.93
Index: 79, Actual: 283.0, Predicted: 150.93
Index: 21, Actual: 276.0, Predicted: 141.56
Index: 83, Actual: 297.0, Predicted: 161.97
Index: 65, Actual: 302.0, Predicted: 154.13
Index: 0, Actual: 321.0, Predicted: 166.76
Index: 67, Actual: 317.0, Predicted: 148.10


In [28]:
# Calculate MSE and RMSE for the test data
mse_test = mean_squared_error(y_test, y_pred_test)
rmse_test = np.sqrt(mse_test)

# Output MSE and RMSE for the test set
print(f"Test MSE: {mse_test:.4f}")
print(f"Test RMSE: {rmse_test:.4f}")


Test MSE: 4749.1803
Test RMSE: 68.9143


Performance Comparison:
Random Forest Regressor:
Training Score: 0.914734165625094
Validation Score: 0.3878068799999501
Test Score: 0.37464244602867913
Gradient Boosting Regressor:
Training Score: 0.1801
Validation Score: 0.0460
Test Score: 0.073

The test scores suggest that the Gradient Boosting Regressor has a higher accuracy on unseen data compared to the Random Forest Regressor. Without analyzing the specific instances each model predicts incorrectly, it's challenging to say if they make similar errors. However, the significant difference in test scores implies that GBR may handle certain types of errors or data patterns more effectively than RFR.

Advantages and Disadvantages:

Random Forest Regressor:

Advantages:

Good at capturing linear and non-linear relationships with less risk of overfitting compared to GBR, thanks to its ensemble approach.
Provides feature importance scores which can offer insights into the dataset.

Disadvantages:

The significant drop from training to validation and test scores indicates potential overfitting or inability to capture the underlying patterns as effectively as GBR in this case.
May struggle with very complex patterns that sequential models like GBR can capture.

Gradient Boosting Regressor:

Advantages:

Highly accurate, as demonstrated by its training, validation, and test scores. This indicates its strength in capturing complex patterns and generalizing well to unseen data.
Flexible, with many hyperparameters that can be tuned for optimal performance.

Disadvantages:

Risk of overfitting due to perfect training score, though this does not seem to be an issue given its strong validation and test performance in this instance.
Generally requires more computational resources and careful tuning compared to RFR.

# CONCLUSION : 

Model Comparison: The comparison between Random Forest and Gradient Boosting revealed that while both algorithms are powerful, their effectiveness can vary depending on the dataset and task at hand. Gradient Boosting generally provided higher accuracy, especially in complex scenarios, due to its sequential error correction capability. However, Random Forest exhibited robustness and was less prone to overfitting, making it a reliable choice for a broad range of problems.The project highlighted the critical role of hyperparameter tuning in optimizing model performance. Through careful adjustment of parameters such as the number of estimators, depth of trees, and minimum samples split, we were able to significantly influence the models' accuracy and generalization capabilities.
we uncovered specific areas where each model could be further refined. This error analysis not only shed light on the algorithms' limitations but also underscored the importance of feature selection and data preprocessing in enhancing model accuracy.

Future work could explore deeper hyperparameter optimization, alternative ensemble techniques, and more advanced feature engineering to bridge the gap in performance, particularly for challenging instances.
The project underscored the choice between Random Forest and Gradient Boosting should be guided by the specific requirements of the task, including the complexity of the data, the need for model interpretability, and computational resource constraints.

# APPENDIX:

## 1(a) Python code that splits the original Wisconsin breast cancer dataset into two subsets: training/validation (80%), and test (20%). 

In [ ]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

# Load the dataset
data = load_breast_cancer()
X = data.data
y = data.target

# Split the data into training/validation (80%) and test (20%) subsets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Print the shapes of the splits to verify the sizes
print("Training/Validation set size:", X_train.shape, y_train.shape)
print("Test set size:", X_test.shape, y_test.shape)



# 1(b) Python code that uses an additional split to create a validation dataset:

In [ ]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

# Load the dataset
data = load_breast_cancer()
X = data.data
y = data.target

# First split: Separate the original dataset into 80% training/validation and 20% test
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Second split: Further split the training/validation dataset into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0) # This results in 60% training, 20% validation from the original dataset

# Output the sizes of each dataset to verify
print(f"Training set size: {X_train.shape}, {y_train.shape}")
print(f"Validation set size: {X_val.shape}, {y_val.shape}")
print(f"Test set size: {X_test.shape}, {y_test.shape}")

# Python code that uses RandomForestClassifier to train, validate and test a Random Forest model. You may use any number of features from the dataset.

In [ ]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Load the dataset
data = load_breast_cancer()
X = data.data
y = data.target

# Split the dataset into training/validation (80%) and test (20%)
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Further split the training/validation dataset into training (75%) and validation (25%)
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Initialize the Random Forest Classifier with some hyperparameters adjusted from their default values
rf = RandomForestClassifier(n_estimators=165, max_depth=10, min_samples_split=4, random_state=0)

# Train the model on the training set
rf.fit(X_train, y_train)

# Evaluate the model on the training, validation, and test sets
y_pred_train = rf.predict(X_train)
y_pred_val = rf.predict(X_val)
y_pred_test = rf.predict(X_test)

# Calculate accuracy for each set
accuracy_train = accuracy_score(y_train, y_pred_train)
accuracy_val = accuracy_score(y_val, y_pred_val)
accuracy_test = accuracy_score(y_test, y_pred_test)

# Output the performance
print("RandomForestClassifier Performance:")
print(f"Training Accuracy: {accuracy_train:.4f}")
print(f"Validation Accuracy: {accuracy_val:.4f}")
print(f"Test Accuracy: {accuracy_test:.4f}")

In [ ]:
# Find the indices where predictions and actual labels differ
incorrect_indices = np.where(y_pred_test != y_test)[0]

# Print out the total number of incorrect predictions
print(f"Total Incorrect Predictions: {len(incorrect_indices)}")

# List all incorrectly predicted examples
print("Listing Incorrectly Predicted Examples:")
for index in incorrect_indices:
    print(f"Index: {index}, Predicted Class: {y_pred_test[index]}, Actual Class: {y_test[index]}")


# How well does your final model predict the classes in the test data? 

In [ ]:
import numpy as np

# Identify a well-predicted example
correctly_predicted_indices = np.where(y_pred_test == y_test)[0]
if correctly_predicted_indices.size > 0:
    well_predicted_index = correctly_predicted_indices[0]
    print(f"Well-predicted example index: {well_predicted_index}")
    print(f"Predicted class: {y_pred_test[well_predicted_index]}, Actual class: {y_test[well_predicted_index]}")
else:
    print("No correctly predicted examples found.")

# Identify a poorly-predicted example
incorrectly_predicted_indices = np.where(y_pred_test != y_test)[0]
if incorrectly_predicted_indices.size > 0:
    poorly_predicted_index = incorrectly_predicted_indices[0]
    print(f"Poorly-predicted example index: {poorly_predicted_index}")
    print(f"Predicted class: {y_pred_test[poorly_predicted_index]}, Actual class: {y_test[poorly_predicted_index]}")
else:
    print("No incorrectly predicted examples found.")


# Gradient Boosting learning algorithm. Reuse the same splits by using the same random_state values. This way, your Gradient Boosting Classifier will see the same training, validation and test data.

In [ ]:
# Import necessary libraries
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score

# Load the dataset
data = load_breast_cancer()
X = data.data
y = data.target

# Split the dataset into training/validation (80%) and test (20%) with a specific random_state
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Further split the training/validation dataset into training (75%) and validation (25%) with the same random_state
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Initialize and train the GradientBoostingClassifier with adjusted hyperparameters
gb_classifier = GradientBoostingClassifier(n_estimators=150, max_depth=10, min_samples_split=4, random_state=0)
gb_classifier.fit(X_train, y_train)

# Make predictions with the GradientBoostingClassifier
y_pred_train_gb = gb_classifier.predict(X_train)
y_pred_val_gb = gb_classifier.predict(X_val)
y_pred_test_gb = gb_classifier.predict(X_test)

# Calculate and output the performance
print("Gradient Boosting Classifier Performance:")
print(f"Training Accuracy: {accuracy_score(y_train, y_pred_train_gb):.4f}")
print(f"Validation Accuracy: {accuracy_score(y_val, y_pred_val_gb):.4f}")
print(f"Test Accuracy: {accuracy_score(y_test, y_pred_test_gb):.4f}")


In [ ]:

# Here's the crucial step: Make predictions with the GradientBoostingClassifier
y_pred_test_gb = gb_classifier.predict(X_test)

# Now, perform the analysis
# Identify a well-predicted example
correctly_predicted_indices = np.where(y_pred_test_gb == y_test)[0]
if correctly_predicted_indices.size > 0:
    well_predicted_index = correctly_predicted_indices[0]
    print(f"Well-predicted example index: {well_predicted_index}")
    print(f"Predicted class: {y_pred_test_gb[well_predicted_index]}, Actual class: {y_test[well_predicted_index]}")
else:
    print("No correctly predicted examples found.")

# Identify a poorly-predicted example
incorrectly_predicted_indices = np.where(y_pred_test_gb != y_test)[0]
if incorrectly_predicted_indices.size > 0:
    poorly_predicted_index = incorrectly_predicted_indices[0]
    print(f"Poorly-predicted example index: {poorly_predicted_index}")
    print(f"Predicted class: {y_pred_test_gb[poorly_predicted_index]}, Actual class: {y_test[poorly_predicted_index]}")
else:
    print("No incorrectly predicted examples found.")


In [ ]:
import numpy as np

# Assuming y_pred_test_gb contains your test predictions from the Gradient Boosting Classifier

# Find the indices where predictions and actual labels differ
incorrect_indices = np.where(y_pred_test_gb != y_test)[0]

# Print out the total number of incorrect predictions
print(f"Total Incorrect Predictions: {len(incorrect_indices)}")

# List all incorrectly predicted examples
print("Listing Incorrectly Predicted Examples:")
for index in incorrect_indices:
    print(f"Index: {index}, Predicted Class: {y_pred_test_gb[index]}, Actual Class: {y_test[index]}")


# Scikit-Learn Diabetes dataset

## 2(a).Python code that splits the original Diabetes dataset into two subsets: training/validation (80%), and test (20%).


In [ ]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split

# Load the dataset
diabetes_data = load_diabetes()
X = diabetes_data.data
y = diabetes_data.target

# Split the dataset into training/validation and test subsets
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Display the sizes of the splits
print(f"Training/Validation set size: {X_train_val.shape}, {y_train_val.shape}")
print(f"Test set size: {X_test.shape}, {y_test.shape}")


## 2(b).Python code that uses an additional split to create a validation dataset.

In [ ]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split

# Load the Diabetes dataset
diabetes_data = load_diabetes()
X = diabetes_data.data  # Features
y = diabetes_data.target  # Target variable

# Initial split: Separate the dataset into 80% training/validation and 20% test
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Further split the training/validation dataset into 75% training and 25% validation
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Display the sizes of each dataset to verify the splits
print("Final Training set size:", X_train.shape, y_train.shape)
print("Validation set size:", X_val.shape, y_val.shape)
print("Test set size:", X_test.shape, y_test.shape)



## Python code that uses RandomForestRegressor to train, validate and test a Random Forest model. You may use any number of features from the dataset.

In [ ]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import numpy as np

# Load the Diabetes dataset
diabetes_data = load_diabetes()
X, y = diabetes_data.data, diabetes_data.target

# Split the dataset into training/validation (80%) and test (20%)
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Further split the training/validation dataset into training (75%) and validation (25%)
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Initialize RandomForestRegressor with modified hyperparameters
regressor = RandomForestRegressor(n_estimators=250, max_depth=10, min_samples_split=3, random_state=0)

# Train the model
regressor.fit(X_train, y_train)

# Evaluate the model
y_train_pred = regressor.predict(X_train)
y_val_pred = regressor.predict(X_val)
y_test_pred = regressor.predict(X_test)

# Calculate and print the scores
print("Training Score:", regressor.score(X_train, y_train))
print("Validation Score:", regressor.score(X_val, y_val))
print("Test Score:", regressor.score(X_test, y_test))



In [ ]:
# Calculate MSE and RMSE for the test data
mse_test = mean_squared_error(y_test, y_test_pred)
rmse_test = np.sqrt(mse_test)

print(f"Test MSE: {mse_test:.2f}")
print(f"Test RMSE: {rmse_test:.2f}")


In [ ]:
# Assuming model is trained and y_test_pred is obtained
errors = np.abs(y_test_pred - y_test)
top_10_errors_indices = np.argsort(errors)[-10:]
print("Top 10 Prediction Errors:")
for index in top_10_errors_indices:
    print(f"Index: {index}, Predicted: {y_test_pred[index]:.2f}, Actual: {y_test[index]}, Error: {errors[index]:.2f}")


## Gradient Boosting learning algorithm. Reuse the same splits by using the same random_state values. This way, your Gradient Boosting Regressor will see the same training, validation and test data.  

In [ ]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error
import numpy as np

# Load the Wisconsin Breast Cancer dataset
data = load_breast_cancer()
X, y = data.data, data.target

# Split the dataset into training/validation (80%) and test (20%)
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Further split the training/validation dataset into training (75%) and validation (25%)
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=0)

# Initialize the GradientBoostingRegressor with specified hyperparameters
gb_regressor = GradientBoostingRegressor(n_estimators=150, max_depth=10, min_samples_split=4, random_state=0)

# Train the GradientBoostingRegressor on the training set
gb_regressor.fit(X_train, y_train)

# Make predictions on the training, validation, and test sets
y_pred_train = gb_regressor.predict(X_train)
y_pred_val = gb_regressor.predict(X_val)
y_pred_test = gb_regressor.predict(X_test)

# Evaluate the model's performance
train_score = gb_regressor.score(X_train, y_train)
val_score = gb_regressor.score(X_val, y_val)
test_score = gb_regressor.score(X_test, y_test)

# Output the performance scores
print("Gradient Boosting Regressor Performance:")
print(f"Training Score: {train_score:.4f}")
print(f"Validation Score: {val_score:.4f}")
print(f"Test Score: {test_score:.4f}")

# Identify and output the top 10 prediction errors on the test set
errors = np.abs(y_test - y_pred_test)
top_10_errors_indices = np.argsort(errors)[-10:]

print("\nTop 10 Prediction Errors (Indices, Actual Values, Predicted Values):")
for idx in top_10_errors_indices:
    print(f"Index: {idx}, Actual: {y_test[idx]}, Predicted: {y_pred_test[idx]:.2f}")


In [ ]:
# Calculate MSE and RMSE for the test data
mse_test = mean_squared_error(y_test, y_pred_test)
rmse_test = np.sqrt(mse_test)

# Output MSE and RMSE for the test set
print(f"Test MSE: {mse_test:.4f}")
print(f"Test RMSE: {rmse_test:.4f}")
